ui/js/graphs.js

```diff
@@ -769,6 +786,77 @@ perf.controller('TestChooserCtrl', function($scope, $modalInstance, $http,
     $scope.testsToAdd = []; // tests in the "tests to add" list
     $scope.selectedTestsToAdd = []; // tests in the "to add" test list that have been selected by the user
 
+    var resolvePromise = $q.defer();
+    var relatedSeries = options.relatedSeries;
+    // if we choose to add all related platforms, firstly we need
+    // to obtain all the platform of this project, then pick out the
+    // specific test by platform and project.
+    var addRelatedPlatforms = function(relatedSeries) {
+        PhSeries.getAllSeries(relatedSeries.projectName,
+            $scope.timeRange, optionCollectionMap).then(
+            function(seriesData) {
+                var platformList = seriesData.platformList;
+                platformList.forEach(function(platform) {
+                    var testList = _.sortBy(_.filter(seriesData.seriesList,
+                        { platform: platform }), 'name');
+                    var temp = _.findWhere(testList, {"name": relatedSeries.name});
+                    // if found something different from the series we already have,
+                    // then we push it into the testsToAdd list.
+                    if (temp !== undefined && temp.signature !== relatedSeries.signature) {
+                        $scope.testsToAdd.push(_.clone(temp));
+                    }
+                });
+            }
+        ).then(function() {
+                resolvePromise.resolve("Done");
+            });
+    };
+    // if we choose to add all related branches, we should
+    // to find out all the branches of specific platform. Then query
+    // the test data one by one.
+    var addRelatedBranches = function(options, relatedSeries) {
+        var branchList = [];
+        thPerformanceBranches.forEach(function(branch) {
+            if (branch !== options.relatedSeries.projectName) {
+                branchList.push(_.findWhere($scope.projects,
+                    {name: branch}));
+            }
+        });
+        branchList.forEach(function(project) {
+            PhSeries.getAllSeries(project.name,
+                $scope.timeRange, optionCollectionMap).then(
+                function(seriesData) {
+                    var testList = _.sortBy(_.filter(seriesData.seriesList,
+                        {platform: relatedSeries.platform}), 'name');
+                    var temp = _.findWhere(testList, {"name": relatedSeries.name});
+                    $scope.testsToAdd.push(_.clone(temp));
+                }
+            ).then(function() {
+                    resolvePromise.resolve("Done");
+                });
```

Comment:
I'm not sure if it's ok to use promise like this(put it into a forEach loop). Maybe you have some more reliable way to work on this? :) @wlach